docs/libvirt-node-networking.md

```diff
@@ -0,0 +1,263 @@
+# * unconfigure $DEV and set the IP on mv4 and manully change the default route
+# using an extra IP via dhcp sounds like the best solution.
+# We can leave all other routes which belong to $DEV untouched.
+# No matter what we do, we don't want any routes modified.
+ip addr add $DEV_SUBNET.13/24 dev mv4 noprefixroute
+
+# normally we already have the netork namespace of libvirt given.
+ip netns add ns1
+
+# create the macvlan link attaching it to the parent host $DEV
+ip link add mv1 link $DEV type macvlan mode bridge
+
+# move the new interface mv1 to the new namespace
+ip link set mv1 netns ns1
+
+# bring the interface up
+# ns1 will later be used by libvirt to attach macvtaps to
+ip netns exec ns1 ip link set dev mv1 up
+
+
+```
+
+After that invocation, we have a host macvlan, with an IP address assigned. The
+CNI dhcp plugin comes with a minimalistic dhcp client, which we can use to do
+the DHCP renewals. An important side note here is, that the minimalistic daemon
+returns the acquired IP alog side with routing rules suggested by the DHCP
+server, but they are not applied, so the host is not modified.
+
+#### Prepare the VM device
+
+When a VM contains a network configuration like this:
+
+```yaml
+kind: VM
+spec:
+  domain:
+    devices:
+      interfaces:
+      - type: HostNetwork
+        mac:
+          address: 00:11:22:33:44:55
+        model:
+          type: virtio
+```
+For every `HostNetwork`, `virt-handler` let's libvirt create a new macvtap
+attached to the one prepared macvlan in the initial setup step in the libvirt
+namespace.  Ee will use the CNI `dhcp` IPAM plugin with a provided or generated
+MAC address. Based on this MAC address and the IPAM plugin invocation we end up
+with the correct IP and MAC pair. The MAC address will then be used by libvirt
+as the mactap address. The IP can be used to add a route, to allow routing
+between the new guest and the host.
+
+Taking the macvlan name and the MAC address, we can now transform the
+HostNetwork entry into a libvirt XML fragment:
+
+```xml
+<interface type='direct'>
+  <mac address="fe:54:00:f7:78:6d"/>
+  <source dev='macvlan1' mode='bridge'/>
+</interface>
+```
+
+When the guest now tries to acquire an IP address with this mac address, it
+will get the same IP address.
+
+```bash
+# Add an extra route which allows accessing the $DEV_IP through the master macvlan.
+# We need one of these routes per device.
+ip route add $DEV_SUBNET.11/32 via $DEV_SUBNET.13 dev mv4
+
+# ping from container to host
+# done by the ipam dhcp client
+ip netns exec ns1 ping $DEV_SUBNET.13
+```
+
+In this scenario, we can keep the IP of our main interface in place, and only
+have to add one additional route per acquired IP for a VM. All existing routes
+can stay untouched. In theory, moving the IP of the main interface to `mv4`
+would also be possible, but that would require extensive route rewrites which
+will likely cause conflicts with other components involved.
+
+### Bridge Scenario (tap)
```

Comment:
+1 on using this solution.

We just need to see how th ebridge is created on the host - do we expect it to be there or do we do it?
Which interfaces get attached to the bridge?
Do we expect the host to have a second NIC in the control network, which is dedicated to KubeVirt usage, or do we share the NIC which is used for K8s?

Reply:
There is no doubt that the bridge solution is the most painless one. However, it requires, that one plans to run KubeVirt already with nested nodes, when deploying k8s or OpenShift. I would prefer to use the bridge solution if possible (someone created bridges already for whatever reason), but use the macvlan solution, as a way to make this work with most of the existing k8s deployments out there.